test/Chainweb/Test/Pact/TTL.hs

```diff
@@ -0,0 +1,244 @@
+import Chainweb.Test.Pact.Utils
+import Chainweb.Test.Utils
+import Chainweb.Time
+import Chainweb.TreeDB
+import Chainweb.Utils
+import Chainweb.Version
+
+
+testVer :: ChainwebVersion
+testVer = Development
+
+tests :: ScheduledTest
+tests =
+    ScheduledTest label $
+    withRocksResource $ \rocksIO ->
+    withPayloadDb $ \pdb ->
+    withBlockHeaderDb rocksIO genblock $ \bhdb ->
+    withTemporaryDir $ \dir ->
+    testGroup label
+        [ withTime $ \iot -> withPact pdb bhdb (testMemPoolAccess iot) dir $ \reqQIO ->
+            testCase "reject-tx-with-badttl" $ testTTL genblock pdb bhdb reqQIO
+        ]
+  where
+    genblock = genesisBlockHeader testVer cid
+    label = "Chainweb.Test.Pact.TTL"
+    cid = someChainId testVer
+
+testTTL
+    :: BlockHeader
+    -> IO (PayloadDb HashMapCas)
+    -> IO (BlockHeaderDb)
+    -> IO (TQueue RequestMsg)
+    -> Assertion
+testTTL genesisBlock iopdb iobhdb rr = do
+    (T3 _ newblock _) <- liftIO $ mineBlock genesisBlock (Nonce 1) iopdb iobhdb rr
+    expectException $ mineBlock newblock (Nonce 2) iopdb iobhdb rr
+  where
+    expectException act = do
+        m <- wrap `catch` h
+        case m of
+          Just msg -> assertBool msg False
+          Nothing -> return ()
+      where
+        h :: SomeException -> IO (Maybe String)
+        h _ = return Nothing
+        wrap = do
+          (T3 _ _ _) <- act
+          return $ Just "Expected a transaction validation failure."
+withPact
+    :: IO (PayloadDb HashMapCas)
+    -> IO BlockHeaderDb
+    -> MemPoolAccess
+    -> IO FilePath
+    -> (IO (TQueue RequestMsg) -> TestTree)
+    -> TestTree
+withPact iopdb iobhdb mempool iodir f =
+    withResource startPact stopPact $ f . fmap snd
+  where
+    startPact = do
+        mv <- newEmptyMVar
+        reqQ <- atomically newTQueue
+        pdb <- iopdb
+        bhdb <- iobhdb
+        dir <- iodir
+        a <- async $ initPactService testVer cid logger reqQ mempool mv
+                                     bhdb pdb (Just dir) Nothing False
+        link a
+        return (a, reqQ)
+
+    stopPact (a, reqQ) = do
+        sendCloseMsg reqQ
+        cancel a
+
+    logger = genericLogger Warn T.putStrLn
+    cid = someChainId testVer
+
+testMemPoolAccess :: IO (Time Integer) -> MemPoolAccess
+testMemPoolAccess iot  = MemPoolAccess
+    { mpaGetBlock = \validate bh hash _header  -> do
+            t <- meinhack bh <$> iot
+            getTestBlock t validate bh hash
+    , mpaSetLastHeader = \_ -> return ()
+    , mpaProcessFork = \_ -> return ()
+    }
+  where
+    meinhack :: BlockHeight -> Time Integer -> Time Integer
+    meinhack b tt =
+      foldl' (flip add) tt (replicate (fromIntegral b) millisecond)
+    getTestBlock txOrigTime validate bHeight@(BlockHeight bh) hash = do
+        akp0 <- stockKey "sender00"
+        kp0 <- mkKeyPairs [akp0]
+        let nonce = T.pack . show @(Time Integer) $ txOrigTime --
+            badttl = 100 * 24 * 60 * 60
```

Comment:
ok that is one case covered -- now cover the other bad cases

  - transaction origination time > block time
  - transaction origination time + ttl < block time

Reply:
Ok. Fair point. I'll get on that.